src/task/schedule.rs

```diff
@@ -275,6 +283,80 @@ pub fn create_task(
     Ok(node)
 }
 
+pub fn create_elf_task(
+    file: &FileHandle,
+    flags: u16,
+    affinity: Option<u32>,
+) -> Result<TaskPointer, SvsmError> {
+    // Map the file's physical memory into the creating task's virtual address space
+    let mapped_addr = valloc(file.size())?;
+    if file
+        .map_view(
+            mapped_addr,
+            0,
+            file.size(),
+            this_cpu().get_current_task(),
+            FileViewPermission::Read,
+        )
+        .is_err()
+    {
+        vfree(mapped_addr, file.size());
+        return Err(SvsmError::Task(TaskError::Alloc));
+    }
+
+    let buf = unsafe { core::slice::from_raw_parts_mut(mapped_addr.as_mut_ptr(), file.size()) };
+
+    let elf = match elf::Elf64File::read(buf) {
+        Ok(elf) => elf,
+        Err(_) => return Err(SvsmError::Module),
+    };
+    let default_base = elf.default_base();
+    let entry = elf.get_entry(default_base);
+
+    let mut task = Task::user_create(entry as usize, 0, flags)?;
+    task.set_affinity(affinity);
+
+    let node = Rc::new(TaskNode {
+        tree_link: RBTreeLink::default(),
+        list_link: Link::default(),
+        task: RefCell::new(task),
+    });
+
+    // Setup the pagetable for the virtual memory ranges described in the file
+    for segment in elf.image_load_segment_iter(default_base) {
+        let vaddr_start = VirtAddr::from(segment.vaddr_range.vaddr_begin);
+        let flags = if segment.flags.contains(elf::Elf64PhdrFlags::EXECUTE) {
+            FileViewPermission::Execute
+        } else if segment.flags.contains(elf::Elf64PhdrFlags::WRITE) {
+            FileViewPermission::Write
+        } else {
+            FileViewPermission::Read
+        };
+
+        let _res = file.map_view(
+            vaddr_start,
+            segment.file_range.offset_begin,
+            segment.file_range.offset_end - segment.file_range.offset_begin,
+            node.clone(),
+            flags,
+        );
+    }
+    {
+        // Ensure the tasklist lock is released before schedule() is called
+        // otherwise the lock will be held when switching to a new context
+        let mut tl = TASKLIST.lock();
+        tl.list().push_front(node.clone());
+        // Allocate any unallocated tasks (including the newly created one)
+        // to the current CPU
+        this_cpu_mut()
+            .runqueue
+            .allocate(this_cpu().get_apic_id(), tl.list());
+    }
+    schedule();
```

Comment:
I'm slightly worried there's a case where we may call schedule without releasing the lock before. Maybe we can add a check